File: app/lib/bands.mjs
/**
 * The tier bands, on the server.
 *
 * This arithmetic is the reader's, and the reader is where it belongs: the
 * cuts are a display decision, and site/spec-reader/app.js owns them. It is
 * copied here rather than imported because app.js is a hundred kilobyte
 * non-module file written for a browser, and importing it would mean making
 * the reader a module to serve a route.
 *
 * A copy is a liability, so app/lib/__tests__/bands.test.mjs extracts the
 * reader's own `tierBand` from the file and holds this one to it over every
 * score a cell can produce. If the two ever disagree the server and the site
 * would call the same passage two different things, which is worse than either
 * being wrong alone.
 */

/** Display tiers, strongest first. The order IS the precedence rule. */
export const TIERS = ["defining", "core", "related"];

/**
 * The tier a score lands in, or null when it is below every tier.
 *
 * Per cell with j judges: defining is score >= 2j+1, clamped to the cell's own
 * maximum so that unanimous core counts as defining on three point data; core
 * is >= 2j; related is >= j+1, at least two judges behind it. A single judge
 * cell has no consensus to demand, so its sole related vote renders at its own
 * weight instead of dying under the multi-judge floor.
 */
export function tierBand(score, judges, maxCell, related = 1) {
  const definingCut = Math.min(2 * judges + 1, maxCell || 2 * judges + 1);
  const relatedCut = judges > 1 ? judges + 1 : related > 0 ? related : 1;
  return score >= definingCut ? "defining"
    : score >= 2 * judges ? "core"
    : score >= relatedCut ? "related"
    : null;
}

/**
 * One cell's passages, each carrying the band it earned.
 *
 * The scale is per cell -- two on the classic rubric, three once any judge in
 * the cell awards a defining -- but the cut follows the passage: a passage
 * scored by fewer judges than its neighbours tops out lower, and measured
 * against the cell's cut it can be structurally incapable of reaching the band
 * it earned.
 */
export function bandCell(passages, related = 1) {
  const maxVerdict = Math.max(
    2, ...passages.flatMap(passage => Object.values(passage.verdicts || {})));

  return passages.map(passage => {
    const verdicts = Object.values(passage.verdicts || {});
    if (!verdicts.length) return { ...passage, band: null };

    const score = verdicts.reduce(
      (total, verdict) => total + (verdict >= 2 ? verdict : verdict === 1 ? related : 0), 0);
    const maxScore = maxVerdict * verdicts.length;
    return {
      ...passage,
      score,
      maxScore,
      band: tierBand(score, Math.max(1, verdicts.length), maxScore, related),
    };
  });
}

/** Whether a band is `floor` or stronger. An unbanded passage is never kept. */
export function atLeastBand(band, floor) {
  if (!band) return false;
  return TIERS.indexOf(band) <= TIERS.indexOf(floor);
}

File: engine/panel/bands.py
"""The reader's tier bands, in Python.

The depth judge is shown the passages a reader sees by default, so the job has to
draw the same line the reader draws. app/lib/bands.mjs already carries the
reader's arithmetic for the server; this carries that file's, and test_bands.py
runs both over every score a cell can hold."""

TIERS = ("defining", "core", "related")

# The bands a reader opens on, before touching a toggle. site/spec-reader/app.js
# owns this decision in its own DEFAULT_BANDS, and test_bands.py reads that line
# out of the file to hold this one to it: the two drifted once, when related was
# made visible by default and this list was not told, and the depth judge went on
# being handed a narrower evidence base than the reader draws.
DEFAULT_BANDS = ("defining", "core", "related")


def tier_band(score, judges, max_cell, related=1):
    """The band a score lands in for a cell of `judges` judges, or None.

    Defining is score >= 2j+1, clamped to the cell's maximum; core is >= 2j;
    related is >= j+1 (at least two judges), or the lone judge's own weight."""
    defining_cut = min(2 * judges + 1, max_cell or 2 * judges + 1)
    related_cut = judges + 1 if judges > 1 else (related if related > 0 else 1)
    if score >= defining_cut:
        return "defining"
    if score >= 2 * judges:
        return "core"
    if score >= related_cut:
        return "related"
    return None


def band_cell(verdicts, related=1):
    """{key: band or None} for one cell, from {key: {judge: verdict}}.

    The cell's scale is the largest verdict any judge gave in it, at least 2; each
    passage is scored on its own judge count."""
    max_verdict = max([2] + [v for judged in verdicts.values() for v in judged.values()])
    out = {}
    for key, judged in verdicts.items():
        values = list(judged.values())
        if not values:
            out[key] = None
            continue
        score = sum(v if v >= 2 else (related if v == 1 else 0) for v in values)
        out[key] = tier_band(score, max(1, len(values)), max_verdict * len(values), related)
    return out


def shown_by_default(verdicts):
    """The keys a reader shows before touching a toggle: every banded passage.

    Defining, core and related alike. Related is drawn thinner, not hidden, so a
    passage one point under the core cut is on the reader's page and belongs in
    the depth judge's evidence with the rest."""
    return [key for key, band in band_cell(verdicts).items() if band in DEFAULT_BANDS]
